aptos-move/move-examples/token_objects/guild/sources/guild.move

```diff
@@ -0,0 +1,309 @@
+/// This module implements the guild token (non-fungible token) including the
+/// functions create the collection and the guild tokens.
+module token_objects::guild {
+    use std::option;
+    use std::signer;
+    use std::string::{Self, String};
+    use std::vector;
+    use aptos_framework::object::{Self, Object};
+    use aptos_token_objects::collection;
+    use aptos_token_objects::token;
+
+    /// The token does not exist
+    const ETOKEN_DOES_NOT_EXIST: u64 = 1;
+    /// The provided signer is not the admin
+    const ENOT_ADMIN: u64 = 2;
+    /// The provided signer is not the owner
+    const ENOT_OWNER: u64 = 3;
+    /// The provided signer is not a guild master
+    const ENOT_GUILD_MASTER: u64 = 4;
+
+    /// The guild token collection name
+    const GUILD_COLLECTION_NAME: vector<u8> = b"Guild Collection Name";
+    /// The guild collection description
+    const GUILD_COLLECTION_DESCRIPTION: vector<u8> = b"Guild Collection Description";
+    /// The guild collection URI
+    const GUILD_COLLECTION_URI: vector<u8> = b"https://guild.collection.uri";
+
+    /// Whitelisted guild masters
+    struct Config has key {
+        whitelist: vector<address>,
+    }
+
+    /// Manager object refs
+    struct Manager has key {
+        /// The extend_ref of the manager object to get its signer
+        extend_ref: object::ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Guild token
+    struct GuildToken has key {
+        /// Used to get the signer of the token
+        extend_ref: object::ExtendRef,
+        /// Member collection name
+        member_collection: String,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Member token
+    struct MemberToken has key {
+        /// Belonging guild
+        guild: Object<GuildToken>,
+        /// Used to burn
+        burn_ref: token::BurnRef,
+    }
+
+    /// Initializes the module, creating the manager object, the guild token collection and the whitelist.
+    fun init_module(sender: &signer) acquires Manager {
+        // Create the manager object to manage the guild collection.
+        create_manager(sender);
```

Comment:
I'm not quite sure how and why a separate resource is needed for management. It also doesn't seem like the manager does anything really and the extend ref is just used to mint tokens. In that case, why not store the extend ref in the collection object? Check this code out as an example: https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/move-examples/rewards_pool/sources/rewards_pool.move#L48

Reply:
First of all, let me clarify what the "manager" is for. Here, the manager is a dummy object to use as an autonomous account instead of a resource account. The admin (i.e., contract owner) creates the manager and lets it own the guild collection so that it's possible to mint guild tokens without the admin involved (in other words, without the admin's signature).

The `Manager` resource containing the `extend_ref` of the manager object is published under the module address so that later, the module locates/accesses the `extend_ref` and obtains the signer of the manger object (to use it in minting guild tokens). Alternatively, `extend_ref` could be stored in `Config,` which is also published under the module address. In this way, we could just reduce the number of resources to publish. Moreover, the `extend_ref` of the manager object could also be stored under the guild collection object address. It seems to me that all of these are possible and where to store the `extend_ref` doesn't seem to matter much. A more important question may be: if the pattern to use a dummy object as an autonomous account (signer) instead of a resource account looks ok or not?